testdata/p4_16_samples_outputs/default-values-operations-1.p4-stderr

```diff
@@ -0,0 +1,18 @@
+[--Wwarn=uninitialized_use] warning: returnTmp may not be completely initialized
+default-values-operations-1.p4(27): [--Wwarn=uninitialized_use] warning: hdr_or_hu may not be completely initialized
```

Comment:
I find these error messages puzzling. Why are the objects not initialized?

Reply:
returnTmp is a structer initialized with {b=false, ....} (which means field H is initialized with {...}). Specification says that default value for header, initialized like that, is to set header to be Invalid. So we have a structure that has a header field set to invalid, that's why warning is given that struct may not be completely initialized. Same goes for hdr_or_hu, that's also a header initialized with {...}.

Reply:
An invalid header is a properly initialized value.
However, I assume that this happens in a later statement.

Reply:
This warning comes from the def-use analysis pass.

Reply:
This warning signals either a bug in this pass or in the def-use pass. We should figure out which one is true and fix it.
Technically the code generated by this pass should have no uninitialized structs or headers - in the -first.p4 file everything seems to be properly initialized. Can you create a small reproduction just for this case?

Reply:
If the bug is in the def-use analysis then we can merge this code.

Reply:
Here is the generated code:
```
H hdr_or_hu;
H hdr_or_hu_0;
rightTmp_0 = { false, hdr_or_hu, (S){i = 1w0,h = hdr_or_hu_0,b = false} };
```
hdr_or_hu is not set to invalid. I suspect that this is the problem.

Reply:
I understand, i overlooked that invalid header is a properly initialized value.
This happens because, whenever a header is initialized with {...}, StructInitializer will just set declaration variable initializer to empty, and leave it to ResetHeader pass to set it to invalid.  Then SimplifyDefUse sees a declaration without an initialization and generates a warning.

Reply:
First implementation had a set invalid header inside StructInitializer, but then after the pass ResetHeaders, there were redudant setInvalid() calls, so i thought this resolved the problem. I will try to go back to that solution and correct it a little to have just one call.

Reply:
So do you have a plan to fix this or should we do it somewhere else?

Reply:
Problem is that ResetHeaders sets to invalid every header that has an empty declaration. Tuples, that have headers, are initialized with headers extracted in declaration variable. I think thats the main problem, if this is done differently, StructInitializer could set headers to invalid without extracting it to declaration and ResetHeaders wouldnt put redudant call. I will think about this solution and try to implement it.

Reply:
the right solution may be to improve the ResetHeaders pass or to run it sooner.

Reply:
Easiest solution would be to put ResetHeaders before SimplifyDefUse, if it can be done.

Reply:
Feel free to try that if it's not too much work. Hopefully it won't affect too many reference outputs either.

Reply:
I found that this problem was already present before my changes. 
This function segment also throws a warning because returnTmp.h is set to invalid and SimplifyDefUse doesn't see this as initialization.
```
S returnTmp;
returnTmp.i = 1;
returnTmp.h.setInvalid();
returnTmp.b = false;
return returnTmp;
```
```
[--Wwarn=uninitialized_use] warning: returnTmp may not be completely initialized
    return returnTmp;
           ^^^^^^^^^
```
Also if we have 
```
H h; 
h.setInvalid()
```
the same warning is thrown.